CMakeLists.txt

```diff
@@ -11,89 +13,90 @@ option_top (VERONA_DOWNLOAD_LLVM "Download cached version of LLVM" ON)
-option(VERONA_EXPENSIVE_SYSTEMATIC_TESTING "Increase the range of seeds covered by systematic testing" OFF)
 option(USE_CRASH_LOGGING "Enable crash logging in the runtime" OFF)
 if (NOT MSVC)
   option(CMAKE_EXPORT_COMPILE_COMMANDS "Export compilation commands" ON)
 endif ()
 
-SET(VERONA_LOCAL_DIST ${CMAKE_BINARY_DIR}/dist CACHE PATH "Default to installing inside build dir" FORCE)
+##########################################################
+# Pass parameters to subbuild.
+##########################################################
+# CMAKE_HOME_DIRECTORY should not be passed down to another project.
+list (APPEND VERONA_DONT_PASS CMAKE_HOME_DIRECTORY)
+# We override the Verona Install prefix to build a local distribution for tests.
+list (APPEND VERONA_DONT_PASS CMAKE_INSTALL_PREFIX)
+
+if (DEFINED VERONA_EXTRA_CMAKE_ARGS)
+  message (WARNING "Do not use VERONA_EXTRA_CMAKE_ARGS! Value ignored")
+endif ()
+set (VERONA_EXTRA_CMAKE_ARGS)
+# Pass parameters to the subproject
+
+# Set up variables to pass arguments to sub project
+get_cmake_property(CACHE_VARS CACHE_VARIABLES)
+foreach (CACHE_VAR ${CACHE_VARS})
+  list (FIND VERONA_DONT_PASS ${CACHE_VAR} INDEX)
+  if (${INDEX} EQUAL -1)
+    list(APPEND VERONA_EXTRA_CMAKE_ARGS
+      -D${CACHE_VAR}=${${CACHE_VAR}}
+    )
+  endif()
+endforeach ()
+
+## Clang-format target
+include(cmake/clangformat.cmake)
 
-# This is to trick the CMake into building LLVM before Verona.
+clangformat_targets()
+
+
+# ExternalProject is used to trick the CMake into building LLVM before Verona.
 # We use two External Projects, so that the LLVM build can complete and install
 # before we start the Verona one.  This is required as we consume CMake
 # artifacts, which are not available otherwise.
 include(ExternalProject)
-if (NOT DEFINED VERONA_LLVM_LOCATION)
-  project(verona CXX)
-  set(DEFAULT_BUILD_TYPE "Release")
-  if (NOT CMAKE_BUILD_TYPE)
-    message(STATUS "Setting build type to '${DEFAULT_BUILD_TYPE}' as none was specified.")
-    set(CMAKE_BUILD_TYPE "${DEFAULT_BUILD_TYPE}" CACHE STRING "Choose the type of build." FORCE)
-  endif()
 
-  set (MLIR_INSTALL ${CMAKE_BINARY_DIR}/$<CONFIG>/mlir)
+set(DEFAULT_BUILD_TYPE "Release")
+if (NOT CMAKE_BUILD_TYPE)
+  message(STATUS "Setting build type to '${DEFAULT_BUILD_TYPE}' as none was specified.")
+  set(CMAKE_BUILD_TYPE "${DEFAULT_BUILD_TYPE}" CACHE STRING "Choose the type of build." FORCE)
+endif()
+
+if (VERONA_DOWNLOAD_LLVM AND DEFINED LLVM_EXTRA_CMAKE_ARGS)
+  message (WARNING "Ignoring LLVM_EXTRA_CMAKE_ARGS as using downloaded LLVM.")
+endif ()
 
+if (VERONA_DOWNLOAD_LLVM AND DEFINED VERONA_EXTERNAL_LLVM_DIR)
+  message (WARNING "Ignoring VERONA_EXTERNAL_LLVM_DIR as using downloaded LLVM.")
+endif ()
+
+##########################################################
+#  Configure MLIR installation
+##########################################################
+if (DEFINED VERONA_EXTERNAL_LLVM_DIR)
```

Comment:
Rather than relying on `VERONA_EXTERNAL_LLVM_DIR` being undefined, I think it would be better to have, in the option group at the top,

```cmake
set (VERONA_EXTERNAL_LLVM_DIR "" CACHE PATH "Path to an existing LLVM installation.")
```

so the option is documented. This defaults to the empty string, which is false-y in a if block.

The option should also be added to `VERONA_DONT_PASS` to avoid getting a "unused variable" warning.

Reply:
Same applies to `LLVM_EXTRA_CMAKE_ARGS` I guess.

Given that `option` is pretty much a shorthand for `set(XXX Default CACHE BOOL "Description")`, I think we could replace `option_top` with something like:
```cmake
macro (set_cache_top VAR DEFAULT TYPE DEFAULT)
  set (${VAR} ${DEFAULT} CACHE ${TYPE} ${DEFAULT})
  list (APPEND VERONA_DONT_PASS ${VAR})
endmacro()
```

Reply:
I have done approximately what you ask.  I don't really understand the CMake variable/cache model, so if you could take a final look that would be great.

Reply:
Sorry that wasn't clear, what I meant was that `option_top` is subsumed by `set_cache_top`, ie.
```cmake
option_top (VERONA_DOWNLOAD_LLVM "..." ON)
```
Can be replaced by
```cmake
set_cache_top(VERONA_DOWNLOAD_LLVM "..." BOOL ON)
```

then we only need one of the two macros.

Reply:
I've refactored slightly, but kept two macros, it make it easier to switch from `option` to `option_top`, if we need to move things around a bit.